source/components/temp_doeclim_component.cpp

```diff
@@ -0,0 +1,528 @@
+    y[2] = temp * -1 * x[2];
+    y[3] = temp * x[0];
+        
+    return;
+}
+    
+//------------------------------------------------------------------------------
+/*! \brief              Calculates element-wise sum of matrix x and y, stores in z
+ *  \param[in] x,y      Assume x,y setup like x = [a,b,c,d] -> x = |a, b|
+ *                                                                 |c, d|
+ *  \param[in] z        Summed 1-d matrix
+ *  \returns            void, summed matrix is stored in z
+ */
+void TempDOECLIMComponent::sum_1d_2x2_matrix(double * x, double * y, double * z) {
+    for(int i=0; i < 4; i++) {
+        z[i] = x[i] + y[i];
+    }
+        
+    return;
+}
+    
+//------------------------------------------------------------------------------
+// documentation is inherited
+void TempDOECLIMComponent::init( Core* coreptr ) {
+    logger.open( getComponentName(), false, Logger::DEBUG );
+    H_LOG( logger, Logger::DEBUG ) << "hello " << getComponentName() << std::endl;
+    
+    tgaveq.set( 0.0, U_DEGC, 0.0 );
+    tgav.set( 0.0, U_DEGC, 0.0 );
+    flux_mixed.set( 0.0, U_W_M2, 0.0 );
+    flux_interior.set( 0.0, U_W_M2, 0.0 );
+    
+    core = coreptr;
+    
+    tgav_constrain.allowInterp( true );
+    tgav_constrain.name = D_TGAV_CONSTRAIN;
+    
+    // Define the doeclim parameters
+    diff.set( 0.55, U_CM2_S );  // default ocean heat diffusivity, cm2/s. value is CDICE default (varname is kappa there).
+    S.set( 3.0, U_DEGC );         // default climate sensitivity, K (varname is t2co in CDICE).
+    alpha.set( 1.0, U_UNITLESS);  // default aerosol scaling, unitless (similar to alpha in CDICE).
+    
+    // Register the data we can provide
+    core->registerCapability( D_GLOBAL_TEMP, getComponentName() );
+    core->registerCapability( D_GLOBAL_TEMPEQ, getComponentName() );
+    core->registerCapability( D_FLUX_MIXED, getComponentName() );
+    core->registerCapability( D_FLUX_INTERIOR, getComponentName() );
+
+    // Register our dependencies
+    core->registerDependency( D_RF_TOTAL, getComponentName() );
+    core->registerDependency( D_RF_BC, getComponentName() );
+    core->registerDependency( D_RF_OC, getComponentName() );
+    core->registerDependency( D_RF_SO2d, getComponentName() );
+    core->registerDependency( D_RF_SO2i, getComponentName() );
+
+}
+
+//------------------------------------------------------------------------------
+// documentation is inherited
+unitval TempDOECLIMComponent::sendMessage( const std::string& message,
+                                    const std::string& datum,
+                                    const message_data info ) throw ( h_exception )
+{
+    unitval returnval;
+    
+    if( message==M_GETDATA ) {          //! Caller is requesting data
+        return getData( datum, info.date );
+        
+    } else if( message==M_SETDATA ) {   //! Caller is requesting to set data
+        H_THROW("TempDOECLIM: sendMessage not yet implemented for message=M_SETDATA.");
```

Comment:
This looks like boilerplate copied from another component.  Surely we can do better than this.  If there are variables that it makes sense to set from the outside, then we should process them here.  For variables that you aren't allowed to set from outside (which might very well be all of them), the error message should be something like "Setting variable XYZ is not allowed."